pointcloud/pointcloud_utils.go

```diff
@@ -76,73 +76,96 @@ func StatisticalOutlierFilter(meanK int, stdDevThresh float64) (func(PointCloud)
+			c2 := make(chan PointAndData, kd.Size())
+			var newWG sync.WaitGroup
 			// fmt.Println("iterating concurrently")
-			// newWG.Add(1)
-			// go func() {
-			// 	defer newWG.Done()
-			// 	distDone := false
-			// 	pointDone := false
-			// 	for {
-			// 		select {
-			// 		case newDist, isDone := <-c1:
-			// 			if isDone {
-			// 				fmt.Println("dist done")
-			// 				distDone = isDone
-			// 				continue
-			// 			}
-			// 			avgDistances = append(avgDistances, newDist)
-			// 		case newPoint, isDone := <-c2:
-			// 			if isDone {
-			// 				fmt.Println("point done")
-			// 				pointDone = isDone
-			// 				continue
-			// 			}
-			// 			points = append(points, newPoint)
-			// 		default:
-			// 			if distDone && pointDone {
-			// 				return
-			// 			}
-			// 			// case isDone := <-c3:
-			// 			// 	if isDone {
-			// 			// 		fmt.Println("done")
-			// 			// 		return
-			// 			// 	}
-			// 		}
-			// 	}
-			// }()
-			pointAsMatrix.IterateConcurrently(3, func(v r3.Vector, d Data) bool {
+			newWG.Add(1)
+			go func() {
+				defer newWG.Done()
+				distDone := false
+				pointDone := false
+				for {
+					select {
+					case newDist, isDone := <-c1:
+						if isDone {
+							// fmt.Println("dist done")
+							distDone = isDone
+							continue
+						}
+						avgDistances = append(avgDistances, newDist)
+					case newPoint, isDone := <-c2:
+						if isDone {
+							// fmt.Println("point done")
+							pointDone = isDone
+							continue
+						}
+						points = append(points, newPoint)
+					default:
+						if distDone && pointDone {
+							return
+						}
+						// case isDone := <-c3:
+						// 	if isDone {
+						// 		fmt.Println("done")
+						// 		return
+						// 	}
+					}
+				}
+			}()
+			pointAsMatrix.IterateConcurrently(4, func(v r3.Vector, d Data) bool {
 				neighbors := kd.KNearestNeighbors(v, meanK, false)
 				sumDist := 0.0
 				for _, p := range neighbors {
 					sumDist += v.Distance(p.P)
 				}
-				iterateMutex.Lock()
-				// c1 <- (sumDist / float64(len(neighbors)))
-				// c2 <- PointAndData{v, d}
+				// iterateMutex.Lock()
+				c1 <- (sumDist / float64(len(neighbors)))
+				c2 <- PointAndData{v, d}
 				// fmt.Println(len(c2))
-				avgDistances = append(avgDistances, sumDist/float64(len(neighbors)))
-				points = append(points, PointAndData{v, d})
-				iterateMutex.Unlock()
+				// avgDistances = append(avgDistances, sumDist/float64(len(neighbors)))
+				// points = append(points, PointAndData{v, d})
+				// iterateMutex.Unlock()
 				return true
 			},
 			)
-			// close(c1)
-			// close(c2)
-			// newWG.Wait()
+			close(c1)
+			close(c2)
+			newWG.Wait()
+			c1Len := len(c1)
+			c2Len := len(c2)
 			// fmt.Println("kd sis:", kd.Size(), "avgdists:", len(c1), "points", len(c2))
-			// close(c1)
-			// close(c2)
-			// for i := range c1 {
-			// 	avgDistances = append(avgDistances, i)
-			// }
+			if float64(kd.Size())*0.97 > float64(c1Len) || float64(kd.Size())*0.97 > float64(c2Len) {
```

Comment:
I decided to go with channels but some points seem to be lost because of concurrency issues / not waiting and closing properly.